### ISSUE TITLE
Alternative ways to specify IO Types

### ISSUE THREAD

**Issue by samreid** [phetsims/tandem#211]
From https://github.com/phetsims/tandem/issues/188#issuecomment-689779651 and discussion with @zepumph and @pixelzoom.

After https://github.com/phetsims/tandem/issues/210 ObjectIO constructor will no longer be used to create wrappers.

This opens up the possibility for 2 different ways to specify IO types that may be better in the long run:

(A) Specify the IO type as an object instead of a `class`.  This means instead of:

```js
class BunnyIO extends AnimalIO{
  static documentation = 'documentation';
  static typeName = '...';
  static fromStateObject(){...}
}
```

It could be declared like so:
```js
const BunnyIO = new IOType({
  documentation: 'documentation',
  typeName: '...'
});
```

or
```js
class BunnyIO extends IOType{
  toStateObject(){...}
}
return new BunnyIO();
```

(B) Use the core type as an IO type.  For instance, instead of BunnyIO declaring a separate type like so:

```js
class BunnyIO extends ObjectIO {

  // @public @override
  static toStateObject( bunny ) { return bunny.toStateObject(); }

  // @public @override
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }

  // @public @override
  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); }
}

setIOTypeFields( BunnyIO, 'BunnyIO', Bunny );
```

Instead these fields would be moved to `Bunny` and it would serve as the IO type.

Constraints and dimensions to consider:
* Subclassing in IO types, also tracking parent IO type automatically
* Parametric IO types
* Convenience and scalability
* Cost to transfer to a new pattern.

**Comment by samreid** [comment]
A point from @zepumph or @pixelzoom, if we were using a different language we would probably identify IO Type as an interface, to decouple the implementation from the API..

**Comment by samreid** [comment]
Note: Work for this issue should be done in patches or branches until Natural Selection has safe SHAs.

**Comment by samreid** [comment]
@zepumph suggested that we could identify APIs for different kids of statefulness.  For instance, data types must implement `toStateObject/fromStateObject` etc.  Not sure if we would need an enumeration or some other way to distinguish between these.  In Java, it would be a different interface or sub-interface.

**Comment by samreid** [comment]
Identifying the desired API for an IO Type has been very helpful:

```js
const IOType = {

  // required
  typeName: '',
  supertype: null,
  // validator keys such as isValidValue, valueType or validValues

  // state
  toStateObject: coreObject => {}, // Any stateful PhET-iO Element
  fromStateObject: stateObject => {}, // Data Types only
  stateToArgsForConstructor: stateObject => {}, // Dynamic Elements only
  applyState: ( o, value ) => { }, // Reference Types only

  // optional
  methods: {},
  events: [],
  methodOrder: [],
  parameterTypes: [],
  validator: {}
};
```

Based on this API, I have sketched out a type `class IOType` which satisfies this API and has a convenient pattern for declaration sites.  I converted several IO Types to use this pattern and it reduces the amount of code in every case significantly.  It also automatically validates, which was a sore spot in our previous implementation.

<details>

```diff
Index: main/tandem/js/types/ObjectIO.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/tandem/js/types/ObjectIO.js	(revision 9bde4c8e22bd80b8a1ace2a44b3eba7917063bd2)
+++ main/tandem/js/types/ObjectIO.js	(date 1599846443126)
@@ -11,302 +11,12 @@
  * @author Andrew Adare (PhET Interactive Simulations)
  * @author Michael Kauzmann (PhET Interactive Simulations)
  */
-
-import validate from '../../../axon/js/validate.js';
-import ValidatorDef from '../../../axon/js/ValidatorDef.js';
-import merge from '../../../phet-core/js/merge.js';
-import PhetioConstants from '../PhetioConstants.js';
+import IOType from './IOType.js';
 import tandemNamespace from '../tandemNamespace.js';
 
-/**
- * @param {Object} phetioObject
- * @param {string} phetioID
- * @constructor
- * @abstract
- */
-class ObjectIO {
-
-  constructor( phetioObject, phetioID ) {
-    assert && assert( phetioObject, 'phetioObject should be truthy' );
-    assert && assert( phetioID, 'phetioID should be truthy' );
-
-    // @public (read-only)
-    this.phetioObject = phetioObject;
-
-    // @public (read-only)
-    this.phetioID = phetioID;
-
-    // Use the validator defined on the constructor to make sure the phetioObject is valid
-    validate( phetioObject, this.constructor.validator );
-  }
-
-  /**
-   * Return the serialized form of the wrapped PhetioObject. Most often this looks like an object literal that holds the
-   * data about the PhetioObject instance.  This can be overridden by subclasses, or types can use ObjectIO type directly
-   * to use this implementation. This implementation should call `validate()` on the parameter with its `validator` field.
-   * @param {*} o
-   * @returns {*}
-   * @public
-   */
-  static toStateObject( o ) {
-    return o;
-  }
-
-  /**
-   * For data-type serialization. Decodes the object from a state into an instance. This can be overridden by subclasses,
-   * or types can use ObjectIO type directly to use this implementation.
-   * @param {*} stateObject - whatever was returned from the toStateObject method
-   * @returns {*} - the deserialized instance of the same type that the toStateObject was provided as a parameter
-   * @public
-   */
-  static fromStateObject( stateObject ) {
-    return stateObject;
-  }
-
-  /**
-   * Map the stateObject to arguments that are passed to the `create` function in PhetioGroup.js (or other
-   * `PhetioDynamicElementContainer` creation functions). Note that other non-serialized args (not dealt with here) may
-   * be supplied as closure variables. This function only needs to be implemented on IO Types that are
-   * phetioDynamicElement: true, such as PhetioGroup or PhetioCapsule elements.
-   * @param {Object} stateObject - from a corresponding`toStateObject` method
-   * @returns {Array.<*>} - the array of arguments to be passed to creation functions like PhetioGroup.create() or PhetioCapsule.getElement().
-   * @public
-   */
-  static stateToArgsForConstructor( stateObject ) {
-    return [];
-  }
-
-  /**
-   * For reference-type serialization. Applies the stateObject value to the object. When setting PhET-iO state, this
-   * function will be called on an instrumented instance to set the stateObject's value to it.
-   * For more understanding, please read https://github.com/phetsims/phet-io/blob/master/doc/phet-io-instrumentation-guide.md#three-types-of-deserialization
-   * @param {PhetioObject} o
-   * @param {Object} value - result from toStateObject
-   * @public
-   */
-  static applyState( o, value ) { }
-
-  /**
-   * Get the supertype of the passed in IO Type.
-   * @example
-   * ObjectIO.getSupertype( SliderIO )
-   * --> NodeIO
-   * ObjectIO.getSupertype( ObjectIO )
-   * --> null
-   *
-   * @param {function(new:ObjectIO)} ioType
-   * @returns {function(new:ObjectIO)|null} - null if `ioType` is ObjectIO, because that is the root of the IO hierarchy
-   * @public
-   */
-  static getSupertype( ioType ) {
-    assert && assert( ObjectIO.isIOType( ioType ), 'IO Type expected' );
-
-    // getPrototypeOf gets the IO Type's parent type, because the prototype is for the parent.
-    const supertype = Object.getPrototypeOf( ioType );
-    return supertype === Object.getPrototypeOf( window.Object ) ? null : supertype;
-  }
-
-  /**
-   * Returns the type hierarchy for the IO Type, from subtypiest to supertypiest
-   * @param {function(new:ObjectIO)} ioType
-   * @public
-   */
-  static getTypeHierarchy( ioType ) {
-    const array = [];
-    while ( ioType !== null ) {
-      array.push( ioType );
-      ioType = ObjectIO.getSupertype( ioType );
-    }
-    return array;
-  }
-
-  /**
-   * Make sure the IO Type has all the required attributes.
-   * @param {function(new:ObjectIO)} ioType - class to check
-   * @public
-   */
-  static validateIOType( ioType ) {
-
-    const typeName = ioType.typeName;
-    assert && assert( !typeName.includes( '.' ), 'Dots should not appear in type names' );
-
-    // Validate that parametric types look as expected
-    if ( typeName.includes( '<' ) ) {
-      assert && assert( Array.isArray( ioType.parameterTypes ), 'angle bracket notation is only used for parametric IO Types that have parameter IO Types' );
-      ioType.parameterTypes.forEach( parameterType => assert && assert( ObjectIO.isIOType( parameterType ), `parameter type should be an IO Type: ${parameterType}` ) );
-    }
-
-    const splitOnParameters = typeName.split( /[<(]/ )[ 0 ];
-    assert && assert( splitOnParameters.endsWith( PhetioConstants.IO_TYPE_SUFFIX ), 'IO Type name must end with IO' );
-    assert && assert( !ioType.prototype.toStateObject, 'toStateObject should be a static method, not prototype one.' );
-    assert && assert( !ioType.prototype.fromStateObject, 'fromStateObject should be a static method, not prototype one.' );
-    assert && assert( !ioType.prototype.applyState, 'applyState should be a static method, not prototype one.' );
-    assert && assert( !ioType.prototype.stateToArgsForConstructor, 'stateToArgsForConstructor should be a static method, not prototype one.' );
-
-    const supertype = ObjectIO.getSupertype( ioType );
-
-    assert && assert( ioType.hasOwnProperty( 'typeName' ), 'typeName is required' );
-
-    // Prevent inheritance of static attributes, which only pertain to each level of the hierarchy, see https://github.com/phetsims/phet-io/issues/1623
-    // Note that parameterTypes do inherit
-    if ( !ioType.hasOwnProperty( 'events' ) ) { ioType.events = []; }
-    if ( !ioType.hasOwnProperty( 'methods' ) ) { ioType.methods = {}; }
-    if ( !ioType.hasOwnProperty( 'documentation' ) ) { ioType.documentation = {}; }
-    if ( !ioType.hasOwnProperty( 'methodOrder' ) ) { ioType.methodOrder = []; }
-
-    // assert that each public method adheres to the expected schema
-    for ( const method in ioType.methods ) {
-      const methodObject = ioType.methods[ method ];
-      if ( typeof methodObject === 'object' ) {
-        assert && assert( ObjectIO.isIOType( methodObject.returnType ), 'return type must be of type IO: ' + methodObject.returnType );
-
-        assert && assert( Array.isArray( methodObject.parameterTypes ),
-          'parameter types must be an array: ' + methodObject.parameterTypes );
-
-        methodObject.parameterTypes.forEach( parameterType => {
-          assert && assert( ObjectIO.isIOType( parameterType ), 'parameter type must be of type IO: ' + parameterType );
-        } );
-
-        assert && assert( typeof methodObject.implementation === 'function',
-          'implementation must be of type function: ' + methodObject.implementation );
-
-        assert && assert( typeof methodObject.documentation === 'string',
-          'documentation must be of type string: ' + methodObject.documentation );
-
-        assert && methodObject.invocableForReadOnlyElements && assert( typeof methodObject.invocableForReadOnlyElements === 'boolean',
-          'invocableForReadOnlyElements must be of type boolean: ' + methodObject.invocableForReadOnlyElements );
-      }
-    }
-
-    assert && assert( ioType.validator, 'validator must be provided' );
-    assert && assert( ioType.documentation, 'documentation must be provided' );
-    assert && ValidatorDef.validateValidator( ioType.validator );
-
-    ioType.hasOwnProperty( 'methodOrder' ) && ioType.methodOrder.forEach( methodName => {
-      assert && assert( ioType.methods[ methodName ], 'methodName not in public methods: ' + methodName );
-    } );
-
-    if ( ioType.hasOwnProperty( 'api' ) ) {
-      assert && assert( ioType.api instanceof Object, 'Object expected for api' );
-      assert && assert( Object.getPrototypeOf( ioType.api ) === Object.prototype, 'no extra prototype allowed on API object' );
-    }
-
-    // Make sure events are not listed again in the ioType
-    const typeHierarchy = ObjectIO.getTypeHierarchy( supertype );
-    assert && ioType.events && ioType.events.forEach( event => {
-      const has = _.some( typeHierarchy, t => t.events.includes( event ) );
-      assert( !has, 'ioType should not declare event that parent also has: ' + event );
-    } );
-  }
-}
-
-/**
- * Checks if type is an IO Type
- * @param {*} type
- * @public
- * @returns {boolean} - true if inherits from ObjectIO or is ObjectIO
- */
-ObjectIO.isIOType = type => type === ObjectIO || type.prototype instanceof ObjectIO;
-
-/**
- * @typeDef {Object} MethodObject
- * @property {string} documentation
- * @property {function()} implementation - the function to execute when this method is called
- * @property {function(new:ObjectIO)} returnType - the return IO Type of the method
- * @property {Array.<function(new:ObjectIO)>} parameterTypes - the parameter IO Types for the method
- * @property {boolean} [invocableForReadOnlyElements=true] - by default, all methods are invocable for all elements.
- *    However, for some read-only elements, certain methods should not be invocable. In that case, they are marked as
- *    invocableForReadOnlyElements: false.
- */
-
-/**
- * The public methods available for this IO Type. Each method is not just a function, but a collection of metadata
- * about the method to be able to serialize parameters and return types and provide better documentation.
- * @type {Object.<string, MethodObject>}
- */
-ObjectIO.methods = {};
-
-/**
- * The list of events that can be emitted at this level (does not include events from supertypes).
- * @type {string[]}
- * @public
- */
-ObjectIO.events = [];
-
-/**
- * Documentation that appears in PhET-iO Studio, supports HTML markup.
- * @public
- */
-ObjectIO.documentation = 'The root of the wrapper object hierarchy.';
-
-/**
- * The name that this TypeIO will have in the public PhET-iO API. In general, this should only be word characters,
- * ending in "IO". Parameteric types are a special subset of TypeIOs that include their parameters in their typeName.
- * If a TypeIO's parameters are other IO Type(s), then they should be included within angle brackets, like
- * "PropertyIO<Boolean>". Some other types use a more custom format for displaying their parameter types, in this case
- * the parameter section of the type name (immediately following "IO") should begin with an open paren, "(". Thus the
- * schema for a typeName could be defined (using regex) as `[A-Z]\w*IO([(<].*){0,1}`. In most cases, parameterized
- * types should also include a `parameterTypes` field on the TypeIO.
- * @type {string}
- * @public
- */
-ObjectIO.typeName = 'ObjectIO';
-
-/**
- * IO Types can specify the order that methods appear in the documentation by putting their names in this list.
- * This list is only for the methods defined at this level in the type hierarchy.
- *
- * After the methodOrder specified, the methods follow in the order declared in the implementation (which
- * isn't necessarily stable).
- * @type {string[]}
- * @public
- */
-ObjectIO.methodOrder = [];
-
-/**
- * For parametric types, they must indicate the types of the parameters here. 0 if nonparametric.
- * @type {function(new:ObjectIO)[]}
- * @public
- */
-ObjectIO.parameterTypes = [];
-
-/**
- * A validator object to be used to validate the core types that IOTypes wrap.
- * @type {ValidatorDef}
- * @public
- */
-ObjectIO.validator = {
-  isValidValue: value => {
-
-    // It could be a number, string, Object, etc.
-    return value !== undefined;
-  }
-};
-
-ObjectIO.validateIOType( ObjectIO );
-
-/**
- * Fills in the boilerplate for static fields of an IO Type.
- * @param {function} ioType - an IO Type
- * @param ioTypeName - classname of IOType
- * @param coreType - the corresponding Core Type
- * @param {Object} [options]
- */
-ObjectIO.setIOTypeFields = ( ioType, ioTypeName, coreType, options ) => {
-
-  options = merge( {
-    documentation: null // {string} if not provided, default is defined below
-  }, options );
-
-  // Fill in static fields in the IO Type.
-  ioType.typeName = ioTypeName;
-  ioType.documentation = options.documentation ||
-                         `IO Type for ${ioTypeName.substring( 0, ioTypeName.length - PhetioConstants.IO_TYPE_SUFFIX.length )}`;
-  ioType.validator = { valueType: coreType };
-
-  // Verify that the IO Type is valid.
-  ObjectIO.validateIOType( ioType );
-};
+const ObjectIO = new IOType( 'ObjectIO', null, {
+  isValidValue: () => true
+} );
 
 tandemNamespace.register( 'ObjectIO', ObjectIO );
 export default ObjectIO;
\ No newline at end of file
Index: main/tandem/js/types/VoidIO.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/tandem/js/types/VoidIO.js	(revision 9bde4c8e22bd80b8a1ace2a44b3eba7917063bd2)
+++ main/tandem/js/types/VoidIO.js	(date 1599847315528)
@@ -8,26 +8,9 @@
  */
 
 import tandemNamespace from '../tandemNamespace.js';
+import IOType from './IOType.js';
 import ObjectIO from './ObjectIO.js';
 
-class VoidIO extends ObjectIO {
-
-  constructor( instance, phetioID ) {
-    assert && assert( false, 'should never be called' );
-    super( instance, phetioID );
-  }
-
-  /**
-   * @returns {undefined}
-   * @public
-   */
-  static toStateObject( object ) {
-    return undefined;
-  }
-}
-
-VoidIO.documentation = 'Type for which there is no instance, usually to mark functions without a return value';
-
 /**
  * We sometimes use VoidIO as a workaround to indicate that an argument is passed in the simulation side, but
  * that it shouldn't be leaked to the PhET-iO client.
@@ -35,9 +18,11 @@
  * @override
  * @public
  */
-VoidIO.validator = { isValidValue: () => true };
-VoidIO.typeName = 'VoidIO';
-ObjectIO.validateIOType( VoidIO );
+const VoidIO = new IOType( 'VoidIO', ObjectIO, {
+  documentation: 'Type for which there is no instance, usually to mark functions without a return value',
+  toStateObject: () => undefined,
+  isValidValue: () => true
+} );
 
 tandemNamespace.register( 'VoidIO', VoidIO );
 export default VoidIO;
\ No newline at end of file
Index: main/tandem/js/LinkedElementIO.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/tandem/js/LinkedElementIO.js	(revision 9bde4c8e22bd80b8a1ace2a44b3eba7917063bd2)
+++ main/tandem/js/LinkedElementIO.js	(date 1599847175854)
@@ -8,36 +8,18 @@
 
 import Tandem from './Tandem.js';
 import tandemNamespace from './tandemNamespace.js';
+import IOType from './types/IOType.js';
 import ObjectIO from './types/ObjectIO.js';
 
-class LinkedElementIO extends ObjectIO {
-
-  /**
-   * @param {LinkedElement} linkedElement
-   * @returns {Object}
-   * @override
-   * @public
-   */
-  static toStateObject( linkedElement ) {
+const LinkedElementIO = new IOType( 'LinkedElementIO', ObjectIO, {
+  documentation: 'A LinkedElement',
+  isValidValue: () => true,
+  toStateObject: linkedElement => {
     assert && Tandem.VALIDATION && assert( linkedElement.element.isPhetioInstrumented(), 'Linked elements must be instrumented' );
     return { elementID: linkedElement.element.tandem.phetioID };
-  }
-
-  /**
-   * @param {Object} stateObject
-   * @returns {Object}
-   * @override
-   * @public
-   */
-  static fromStateObject( stateObject ) {
-    return {};
-  }
-}
-
-LinkedElementIO.documentation = 'A LinkedElement';
-LinkedElementIO.validator = { isValidValue: () => true };
-LinkedElementIO.typeName = 'LinkedElementIO';
-ObjectIO.validateIOType( LinkedElementIO );
+  },
+  fromStateObject: stateObject => {}
+} );
 
 tandemNamespace.register( 'LinkedElementIO', LinkedElementIO );
 export default LinkedElementIO;
\ No newline at end of file
Index: main/phet-io/doc/phet-io-instrumentation-guide.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/phet-io/doc/phet-io-instrumentation-guide.md	(revision 46a3986178ad25b133afa31259b7fbe3182ec2b5)
+++ main/phet-io/doc/phet-io-instrumentation-guide.md	(date 1599843809645)
@@ -596,7 +596,7 @@
 Property value. The build-in deserialization method will vary based on the type of deserialization taking place (see 
 below).
 
-Also note that the since work completed in https://github.com/phetsims/phet-io/issues/1398, PhET-iO interframe 
+Also, note that the since work completed in https://github.com/phetsims/phet-io/issues/1398, PhET-iO interframe 
 communications run on structured cloning (via `PostMessage`), and not just JSON strings. This means that a failure to 
 implement a proper `toStateObject` in the IO Type will result in a hard fail when trying to send instances of that type 
 to the wrapper frame. The error will likely look something like this: "Something in message was uncloneable when sent 
Index: main/phet-core/js/EnumerationIO.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/phet-core/js/EnumerationIO.js	(revision 3c2aaa3b15d2b877997edcab93c35f46aa2f25e8)
+++ main/phet-core/js/EnumerationIO.js	(date 1599847021446)
@@ -6,6 +6,7 @@
  * @author Sam Reid (PhET Interactive Simulations)
  */
 
+import IOType from '../../tandem/js/types/IOType.js';
 import ObjectIO from '../../tandem/js/types/ObjectIO.js';
 import Enumeration from './Enumeration.js';
 import phetCore from './phetCore.js';
@@ -34,53 +35,20 @@
 /**
  * Creates a Enumeration IOType
  * @param {Enumeration} enumeration
- * @returns {function(new:ObjectIO)}
+ * @returns {IOType}
  */
 const create = enumeration => {
-
-  class EnumerationIOImpl extends ObjectIO {
-    constructor( a, b ) {
-      assert && assert( false, 'This constructor is not called, because enumeration values, like primitives, are never wrapped.' );
-      super( a, b );
-    }
-
-    /**
-     * Encodes an Enumeration value to a string.
-     * @param {Object} value from an Enumeration instance
-     * @returns {Object} - a state object
-     * @override
-     * @public
-     */
-    static toStateObject( value ) {
-      return toStateObjectImpl( value );
-    }
-
-    /**
-     * Decodes a string into an Enumeration value.
-     * @param {string} stateObject
-     * @returns {Object}
-     * @override
-     * @public
-     */
-    static fromStateObject( stateObject ) {
-      assert && assert( typeof stateObject === 'string', 'unsupported EnumerationIO value type, expected string' );
-      assert && assert( enumeration.KEYS.indexOf( stateObject ) >= 0, `Unrecognized value: ${stateObject}` );
-      return enumeration[ stateObject ];
-    }
-  }
-
   const toStateObjectImpl = v => v.name;
   const valueNames = enumeration.VALUES.map( toStateObjectImpl );
 
   // Enumeration supports additional documentation, so the values can be described.
   const additionalDocs = enumeration.phetioDocumentation ? ` ${enumeration.phetioDocumentation}` : '';
-
-  EnumerationIOImpl.validator = { valueType: enumeration };
-  EnumerationIOImpl.documentation = `Possible values: ${valueNames}.${additionalDocs}`;
-  EnumerationIOImpl.typeName = `EnumerationIO(${valueNames.join( '|' )})`;
-  ObjectIO.validateIOType( EnumerationIOImpl );
-
-  return EnumerationIOImpl;
+  return new IOType( `EnumerationIO(${valueNames.join( '|' )})`, ObjectIO, {
+    valueType: enumeration,
+    documentation: `Possible values: ${valueNames}.${additionalDocs}`,
+    toStateObject: coreObject => coreObject.name,
+    fromStateObject: stateObject => enumeration[ stateObject ],
+  } );
 };
 
 phetCore.register( 'EnumerationIO', EnumerationIO );
Index: main/tandem/js/types/IOType.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/tandem/js/types/IOType.js	(date 1599846542648)
+++ main/tandem/js/types/IOType.js	(date 1599846542648)
@@ -0,0 +1,54 @@
+// Copyright 2020, University of Colorado Boulder
+
+import validate from '../../../axon/js/validate.js';
+import ValidatorDef from '../../../axon/js/ValidatorDef.js';
+import merge from '../../../phet-core/js/merge.js';
+import PhetioConstants from '../PhetioConstants.js';
+import tandemNamespace from '../tandemNamespace.js';
+
+// constants
+const VALIDATE_OPTIONS_FALSE = { validateValidator: false };
+
+class IOType {
+  constructor( ioTypeName, supertype, config ) {
+    config = merge( {
+
+      // Requires validator, such as isValidValue | valueType | validValues
+
+      // Optional
+      methods: {},
+      events: [],
+      methodOrder: [],
+      parameterTypes: [],
+      documentation: `IO Type for ${ioTypeName.substring( 0, ioTypeName.length - PhetioConstants.IO_TYPE_SUFFIX.length )}`,
+
+      // state
+      toStateObject: coreObject => coreObject, // Any stateful PhET-iO Element
+      fromStateObject: stateObject => stateObject, // Data Types only
+      stateToArgsForConstructor: stateObject => stateObject, // Dynamic Elements only
+      applyState: ( coreObject, value ) => { } // Reference Types only
+    }, config );
+
+    assert && assert( ValidatorDef.containsValidatorKey( config ), 'Validator is required' );
+
+    // @public (read-only)
+    this.typeName = config.typeName;
+    this.methods = config.methods;
+    this.events = config.events;
+    this.methodOrder = config.methodOrder;
+    this.parameterTypes = config.parameterTypes;
+    this.toStateObject = coreObject => {
+      validate( coreObject, config, VALIDATE_OPTIONS_FALSE );
+      return config.toStateObject( coreObject );
+    };
+    this.fromStateObject = config.fromStateObject;
+    this.stateToArgsForConstructor = config.stateToArgsForConstructor;
+    this.applyState = ( coreObject, value ) => {
+      validate( coreObject, config, VALIDATE_OPTIONS_FALSE );
+      config.applyState( coreObject, value );
+    };
+  }
+}
+
+tandemNamespace.register( 'IOType', IOType );
+export default IOType;
\ No newline at end of file

```
</details>

For example, this saves 32 lines of code in EnumerationIO, 15 lines of code in VoidIO and 18 lines of code in LinkedElementIO.

I'm excited about this pattern and think it is the right way to go.  However, I am reluctant to start making these changes in our codebase until Natural Selection has SHAs next week.

**Comment by samreid** [comment]
I added the proposal for IOType.js and ObjectIO2.js (so named so it doesn't interfere with the existing ObjectIO.js).  These changes allow BunnyIO to be changed from:

```js
class BunnyIO extends ObjectIO {

  /**
   * Serializes a Bunny instance.
   * @param bunny
   * @returns {Object}
   * @public @override
   */
  static toStateObject( bunny ) {
    assert && assert( bunny instanceof Bunny, 'invalid Bunny' );
    return bunny.toStateObject();
  }

  /**
   * Creates the args that BunnyGroup uses to instantiate a Bunny.
   * @param {Object} state
   * @returns {Object[]}
   * @public
   * @override
   */
  static stateToArgsForConstructor( state ) {
    return Bunny.stateToArgsForConstructor( state );
  }

  /**
   * Restores Bunny state after instantiation.
   * @param {Bunny} bunny
   * @param {Object} stateObject
   * @public
   * @override
   */
  static applyState( bunny, stateObject ) {
    assert && assert( bunny instanceof Bunny, 'invalid Bunny' );
    bunny.applyState( stateObject );
  }
}

ObjectIO.setIOTypeFields( BunnyIO, 'BunnyIO', Bunny );

// @public
Bunny.BunnyIO = BunnyIO;
```

To:

```js
// @public
Bunny.BunnyIO = new IOType( 'BunnyIO', ObjectIO, {
  valueType: Bunny,
  toStateObject: bunny => bunny.toStateObject(),
  stateToArgsForConstructor: Bunny.stateToArgsForConstructor,
  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
} );
```

**Comment by samreid** [comment]
Above I referenced issues that will be partially or completely addressed by this proposal.

**Comment by pixelzoom** [comment]
This is looking nice. I'll probably have some more input when I use it in context. But for now, here are a few thoughts...

* Since `supertype` can be `null`, should it be moved into `config`?   I don't feel too strongly about this if `ObjectIO` is the only case where it will be `null`.

* Would it be better to have `supertype` default to `ObjectIO`?  I realize that's a little problematic because `ObjectIO2` is defined using `new IOType`.  I don't feel too strongly about this, and maybe it's even better to have `ObjectIO` explicitly shown as the supertype.  Compare:

```js
// @public
Bunny.BunnyIO = new IOType( 'BunnyIO', ObjectIO, {
  valueType: Bunny,
  toStateObject: bunny => bunny.toStateObject(),
  stateToArgsForConstructor: Bunny.stateToArgsForConstructor,
  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
} );
```

vs:

```js
// @public
Bunny.BunnyIO = new IOType( 'BunnyIO', {
  valueType: Bunny,
  toStateObject: bunny => bunny.toStateObject(),
  stateToArgsForConstructor: Bunny.stateToArgsForConstructor,
  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
} );
```

* If `super type` is `null`, then `toStateObject`, `fromStateObject`, `stateToArgsForConstructor`, and `applyState` all become required, since there needs to be default implementations at the root of the IO Type hierarchy.  If I understand that correctly, then there should be something that verifies this (assertion? validator? something like `ObjectIO.validateIOType`?)

**Comment by pixelzoom** [comment]
I guess another thing I should comment on is whether I think this approach is better than the current approach (see examples in https://github.com/phetsims/tandem/issues/211#issuecomment-691504711).   It's certainly more concise.  But it creates sort of an alternate form of type hierarchy, where you need things like `getTypeHierarchy`. And that seems more complicated and less straightforward.  If we had static constants would this new approach be as attractive?  If I had to choose between conciseness and a standard type hierarchy, I'd choose the latter.

**Comment by samreid** [comment]
Thanks, I agree that `supertype` should be an option that defaults to `ObjectIO` and I added that in the first commit.

I also added assertions that the root IO type supplies all state functions, in the 2nd commit.

> I guess another thing I should comment on is whether I think this approach is better than the current approach (see examples in https://github.com/phetsims/tandem/issues/211#issuecomment-691504711).   It's certainly more concise.  But it creates sort of an alternate form of type hierarchy, where you need things like `getTypeHierarchy`. And that seems more complicated and less straightforward.  If we had static constants would this new approach be as attractive?  If I had to choose between conciseness and a standard type hierarchy, I'd choose the latter. 

The current paradigm also requires a thing called `getTypeHierarchy` which is currently in production.  https://github.com/phetsims/tandem/blob/9bde4c8e22bd80b8a1ace2a44b3eba7917063bd2/js/types/ObjectIO.js#L109-L121 And the implementation of `ObjectIO.getSupertype` is based on `Object.getPrototypeOf( ioType )` to identify parent types.  We will be able to eliminate `ObjectIO.getSupertype` with this proposal.

We are creating a tree of "types", but that doesn't necessarily mean that `class extends` is the best fit (since they aren't actual JS types). I think we should picture the IO Types more like data values rather than creating a new type hierarchy.  See the details which compare the three possible approaches:

<details>

```js

// Strategy A: Tree with one type and data values
class TreeElement {
  constructor( parent, name, computeData ) {
    this.parent = parent;
    this.name = name;
    this.computeData = computeData;
  }
}

const root = new TreeElement( null, 'root', () => 123 );
const child = new TreeElement( root, 'child', () => 456 );

// Strategy B: Tree with one type per level and static data

class Root {
  static name = 'root';

  static computeData() { return 123;}
}

class Child extends Root {
  static name = 'child';

  static computeData() {return 456;}
}

// Strategy C: Tree with one type per level and instance data
class Root {
  constructor( name ) {
    this.name = name;
  }
  computeData(){
    return 123;
  }
}

class Child extends Root {
  constructor( parent, name, computeData ) {
    super( name, computeData );
    this.parent = parent;
  }

  computeData() {
    return 456;
  }
}
```

</details>

Aside from the improved conciseness, this proposal addresses several of the issues listed before this comment: https://github.com/phetsims/tandem/issues/211#issuecomment-691507751

It also eliminates the need for `validateIOSubtype` and `setIOTypeFields`.

> If we had static constants would this new approach be as attractive?

Can you elaborate on this part?  I'm not sure what this means.

**Comment by pixelzoom** [comment]
> >  If we had static constants would this new approach be as attractive?
> Can you elaborate on this part? I'm not sure what this means.

I'm referring your comment https://github.com/phetsims/tandem/issues/188#issuecomment-687643208, which was:

> Also, how much of this will we want to revisit once we can use static properties, like so?

```js
class BunnyIO extends ObjectIO {
  static documentation = '...';
}
```

**Comment by samreid** [comment]
Thanks, I understand. The proposed approach seems superior to using `class extends` and static properties, so I don't think it would need to be revisited once we support static properties.

**Comment by pixelzoom** [comment]
So the next questions is...  When can/should this new approach be rolled into Natural Selection?  I have a deadline to take shas for 1.2 on 9/17, so I'm reluctant to change now.  On the other hand, Natural Selection is (I think?) the only sim currently using the `setIOTypeFields` approach.  So it might be nice to kill off that approach asap, so that we don't have yet-another approach to maintain (in the release branch).  I'm getting very concerned about how many of my sims are using out-dated iO patterns, and the fact that I've been allocated no time to bring them forward to latest patterns.

**Comment by samreid** [comment]
I reviewed the proposal with @zepumph today and he indicated the following:

* This seems like a reasonable approach.  It may not be the final solution but it is a good step in the right direction.  In the future, we may wish to move this even more to the core type.  But having a well-defined API for the IO Type makes that feasible.  This approach also streamlines the common use case for IO Types.
* There is concern about https://github.com/phetsims/scenery/issues/1046, that is, moving NodeIO and TextIO attributes somewhere.  Eventually they will be moved to Node and Text, but to get this working for Thursday, I would likely move them somewhere else.
* @zepumph had other thoughts about tracking metadata which I didn't follow, but we will discuss that more in the future.

**Comment by pixelzoom** [comment]
> So the next questions is... When can/should this new approach be rolled into Natural Selection? I have a deadline to take shas for 1.2 on 9/17, so I'm reluctant to change now. 

@samreid and I discussed on Slack.   We'll defer rolling this in NS until after shas are taken for 1.2.

**Comment by pixelzoom** [comment]
Self unassigning until this is ready for integration into natural-selection and ph-scale.

**Comment by samreid** [comment]
I'm planning to migrate all IO types to this new pattern in one of the first pushes.  OK if I do natural selection and ph-scale and reassign to you for review?

**Comment by pixelzoom** [comment]
Yes, but that's going to happen after I take shas for natural-selection 1.2, correct?

**Comment by samreid** [comment]
Yes

**Comment by samreid** [comment]
I started on this yesterday but will not push to master until NS SHAs are taken.

**Comment by samreid** [comment]
NS SHAs have been taken.  Also, @zepumph reported that EFAC has been cherry picked to the branch, and it is now safe to proceed in master.

**Comment by samreid** [comment]
As I described at today's PhET-iO meeting today, I'll make incremental commits to move toward completion of this and related issues.  It is a lot of work and would be unwieldy to do in branches or as one single commit.

The first commit will remove ObjectIO and port all IO Types to use IOType.  These local tests are passing:
* PhET brand aqua fuzzing
* PhET-iO brand aqua fuzzing
* Natural Selection, Charges and Fields, Energy Forms and Changes passing state and studio fuzzing

I am expecting some things on CT to break, but I will continue to monitor and make updates.  Next steps include:

* Moving IO Types into core files
* Delegate to core types
* Addressing TODOs marked with https://github.com/phetsims/tandem/issues/211
* Convert to arrow functions
* Address CT issues.

**Comment by samreid** [comment]
Duplicating prior comment due to so many hidden comments:

As I described at today's PhET-iO meeting today, I'll make incremental commits to move toward completion of this and related issues.  It is a lot of work and would be unwieldy to do in branches or as one single commit (155 files across 35+ repos).

The first commit will remove ObjectIO and port all IO Types to use IOType.  These local tests are passing:
* Lint everything
* PhET brand aqua fuzzing
* PhET-iO brand aqua fuzzing
* Natural Selection, Charges and Fields, Energy Forms and Changes passing state and studio fuzzing
* Precommit hooks all passing (lint + unit tests)

I am expecting some things on CT to break, but I will continue to monitor and make updates.  Next steps include:

* Moving IO Types into core files
* Delegate to core types
* Addressing TODOs marked with https://github.com/phetsims/tandem/issues/211
* Convert to arrow functions
* Address CT issues.

**Comment by samreid** [comment]
This work is basically stable, and I'll put further work on hold until I can coordinate with @zepumph more.  The reason it is marked as blocks publication is until review is complete and until we decide how QA should test these nontrivial changes (isolated test vs rolling testing into an upcoming dev or RC test).

**Comment by samreid** [comment]
Here's a patch of half-baked ideas trying to get rid of IOType as much as possible or integrate with the core type:

<details>

```diff

Index: main/tandem/js/types/IOType.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/tandem/js/types/IOType.js	(revision ed99942a75bb5afc71d79c640084d22c2b73c2ed)
+++ main/tandem/js/types/IOType.js	(date 1600912799751)
@@ -111,6 +111,9 @@
     this.documentation = config.documentation;
     this.methods = config.methods;
     this.events = config.events;
+
+    this.valueType = config.valueType;
+
     this.methodOrder = config.methodOrder;
     this.parameterTypes = config.parameterTypes;
     this.validator = _.pick( config, ValidatorDef.VALIDATOR_KEYS );
@@ -199,14 +202,17 @@
 
 // @private
 ObjectIO = new IOType( 'ObjectIO', {
-  isValidValue: () => true,
-  supertype: null,
-  documentation: 'The root of the IO Type hierarchy',
-  toStateObject: coreObject => coreObject,
-  fromStateObject: stateObject => stateObject,
-  stateToArgsForConstructor: stateObject => [],
-  applyState: ( coreObject, stateObject ) => { }
-} );
+    isValidValue: () => true,
+    supertype: null,
+    documentation: 'The root of the IO Type hierarchy',
+    toStateObject: coreObject => coreObject.toStateObject ? coreObject.toStateObject() : coreObject,
+    fromStateObject: stateObject => this.valueType && this.valueType.fromStateObject ? this.valueType.fromStateObject( stateObject ) : stateObject,
+    stateToArgsForConstructor: stateObject => this.valueType && this.valueType.stateToArgsForConstructor ? this.valueType.stateToArgsForConstructor(stateObject):[],
+    applyState: ( coreObject, stateObject ) => {
+      coreObject.applyState ? coreObject.applyState( stateObject ) : null;
+    }
+  }
+);
 
 IOType.ObjectIO = ObjectIO;
 
Index: main/natural-selection/js/common/model/BunnyGroup.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/natural-selection/js/common/model/BunnyGroup.js	(revision 4d4ea8f07e1fd9cb4769a4eafbbc1a67a3bd3927)
+++ main/natural-selection/js/common/model/BunnyGroup.js	(date 1600913102697)
@@ -36,7 +36,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: PhetioGroupIO( Bunny.BunnyIO ),
+      phetioType: PhetioGroupIO( Bunny ),
       phetioDocumentation: 'manages dynamic PhET-iO elements of type Bunny, including live and dead bunnies'
     }, options );
 
Index: main/natural-selection/js/common/model/BunnyArray.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/natural-selection/js/common/model/BunnyArray.js	(revision 4d4ea8f07e1fd9cb4769a4eafbbc1a67a3bd3927)
+++ main/natural-selection/js/common/model/BunnyArray.js	(date 1600913102691)
@@ -27,7 +27,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: ObservableArray.ObservableArrayIO( ReferenceIO( Bunny.BunnyIO ) ),
+      phetioType: ObservableArray.ObservableArrayIO( ReferenceIO( Bunny ) ),
       phetioState: false
     }, options );
 
Index: main/natural-selection/js/common/model/Bunny.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/natural-selection/js/common/model/Bunny.js	(revision 4d4ea8f07e1fd9cb4769a4eafbbc1a67a3bd3927)
+++ main/natural-selection/js/common/model/Bunny.js	(date 1600913102687)
@@ -65,7 +65,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: Bunny.BunnyIO,
+      phetioType: Bunny,
       phetioDynamicElement: true
     }, options );
 
@@ -342,8 +342,8 @@
     return {
 
       // Even though father and mother are stateful, we need a reference to them.
-      father: NullableIO( ReferenceIO( Bunny.BunnyIO ) ).toStateObject( this.father ),
-      mother: NullableIO( ReferenceIO( Bunny.BunnyIO ) ).toStateObject( this.mother ),
+      father: NullableIO( ReferenceIO( Bunny ) ).toStateObject( this.father ),
+      mother: NullableIO( ReferenceIO( Bunny ) ).toStateObject( this.mother ),
 
       generation: NumberIO.toStateObject( this.generation ),
       isAlive: BooleanIO.toStateObject( this.isAlive ),
@@ -385,8 +385,8 @@
     required( stateObject );
 
     // public fields
-    this.father = required( NullableIO( ReferenceIO( Bunny.BunnyIO ) ).fromStateObject( stateObject.father ) );
-    this.mother = required( NullableIO( ReferenceIO( Bunny.BunnyIO ) ).fromStateObject( stateObject.mother ) );
+    this.father = required( NullableIO( ReferenceIO( Bunny ) ).fromStateObject( stateObject.father ) );
+    this.mother = required( NullableIO( ReferenceIO( Bunny ) ).fromStateObject( stateObject.mother ) );
     this.generation = required( NumberIO.fromStateObject( stateObject.generation ) );
     this.isAlive = required( BooleanIO.fromStateObject( stateObject.isAlive ) );
     this.age = required( NumberIO.fromStateObject( stateObject.age ) );
@@ -438,12 +438,9 @@
  * as described in the Serialization section of
  * https://github.com/phetsims/phet-io/blob/master/doc/phet-io-instrumentation-guide.md#serialization
  */
-Bunny.BunnyIO = new IOType( 'BunnyIO', {
-  valueType: Bunny,
-  toStateObject: bunny => bunny.toStateObject(),
-  stateToArgsForConstructor: Bunny.stateToArgsForConstructor,
-  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
-} );
+// Bunny.BunnyIO = new IOType( 'BunnyIO', {
+//   valueType: Bunny
+// } );
 
 naturalSelection.register( 'Bunny', Bunny );
 export default Bunny;
\ No newline at end of file
Index: main/natural-selection/js/common/model/SelectedBunnyProperty.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- main/natural-selection/js/common/model/SelectedBunnyProperty.js	(revision 4d4ea8f07e1fd9cb4769a4eafbbc1a67a3bd3927)
+++ main/natural-selection/js/common/model/SelectedBunnyProperty.js	(date 1600913102693)
@@ -33,7 +33,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: PropertyIO( NullableIO( ReferenceIO( Bunny.BunnyIO ) ) ),
+      phetioType: PropertyIO( NullableIO( ReferenceIO( Bunny ) ) ),
       phetioDocumentation: 'the selected bunny, null if no bunny is selected'
     }, options );
 

```

</details>

**Comment by samreid** [comment]
If we automatically leverage `valueType` in ObjectIO like so:

```js
ObjectIO = new IOType( 'ObjectIO', {
  isValidValue: () => true,
  supertype: null,
  documentation: 'The root of the IO Type hierarchy',
  toStateObject: function( coreObject ) {
    return coreObject.toStateObject ? coreObject.toStateObject() : coreObject;
  },
  fromStateObject: function( stateObject ) {
    return this.valueType && this.valueType.fromStateObject ? this.valueType.fromStateObject( stateObject ) : stateObject;
  },
  stateToArgsForConstructor: function( stateObject ) {
    return this.valueType && this.valueType.stateToArgsForConstructor ? this.valueType.stateToArgsForConstructor( stateObject ) : [];
  },
  applyState: function( coreObject, stateObject ) {
    coreObject.applyState ? coreObject.applyState( stateObject ) : null;
  }
} );
```

Then 7 IO Type declarations in Natural Selection change from looking like this:

```js
Wolf.WolfIO = new IOType( 'WolfIO', {
  valueType: Wolf,
  toStateObject: wolf => wolf.toStateObject(),
  stateToArgsForConstructor: Wolf.stateToArgsForConstructor,
  applyState: ( wolf, stateObject ) => wolf.applyState( stateObject )
} );
```

to looking like this:

```js
Wolf.WolfIO = new IOType( 'WolfIO', {
  valueType: Wolf
} );
```

Could we eliminate those altogether?  Perhaps places that expect an IOType could be changed to take a core type or an IO Type.  If given a core type, they would run something like this:

```js
if (type instanceof IOType){
  return type;
} else{
  type.IOType = type.IOType || new IOType(type.phetioTypeName,{valueType: type});
  return type.IOType;
}
```

Then usage sites could read like:

```js
phetioType: ObservableArray.ObservableArrayIO( ReferenceIO( Bunny ) ), // instead of Bunny.BunnyIO

  methods: {
    getSelectedBunny: {
      returnType: Bunny, // Instead of Bunny.BunnyIO
      parameterTypes: [],
      implementation: function() {
        this.getSelectedBunny();
      },
      documentation: 'Gets the selected Bunny'
    }
  }
```

Maybe even `phetioType: Bunny.BunnyIO,` in Bunny.js could be eliminated and automatically inferred.  But I wonder if these proposals to eliminate these altogether are too clever in attempt to replace a process that is already clear and concise?

**Comment by pixelzoom** [comment]
It does seem like this is going in the direction of eliminating IO Types altogether, and having Core Types implement a specific interface. (Too bad JS doesn't have interfaces.)

**Comment by zepumph** [comment]
>Could we eliminate those altogether? Perhaps places that expect an IOType could be changed to take a core type or an IO Type. If given a core type, they would run something like this:

@samreid and I think that it is possible that this could be an improvement, but we don't think we should prematurely investigate. If in the future we find ourselves annoyed at this much boilerplate, then we can reevaluate in a new issue:

`Wolf.WolfIO = new IOType( 'WolfIO', { valueType: Wolf } );`

**Comment by samreid** [comment]
> It does seem like this is going in the direction of eliminating IO Types altogether, and having Core Types implement a specific interface. (Too bad JS doesn't have interfaces.)

Even in languages with interfaces, how is it possible for a type object to implement an interface?  For instance, if there is a bunny named `myBunny = new Bunny()`, we would want `Bunny` to implement the interface rather than the usual thing of wanting `myBunny` to implement the interface.  Keep in mind the interface we are aiming for is roughly described in https://github.com/phetsims/tandem/issues/211#issuecomment-691240527

**Comment by zepumph** [comment]
My individual review of this issue:
* Please go back and look through every open issue linked above. Most were marked as "on hold" until this work was done. I feel confident in the pattern outlined by `Wolf.WolfIO = new IOType( 'WolfIO', { valueType: Wolf } );` such that those can now be dealt with or closed
* I think there are many IO types that still need to be inlined (we discussed this this morning)
* I still see 22 links to this issue in the code (REVIEW and TODO)
* It would be good to double check that everything you mentioned in https://github.com/phetsims/tandem/issues/211#issuecomment-694622549 is complete.

Everything else seems to be in side issues. @samreid please go through the above, at least briefly, before unmarking blocks-publication. I think this is likely ready now, but want to double check that this doesn't spur more thoughts.

**Comment by samreid** [comment]
I reviewed the linked issues, and was able to close several of them.  Most remaining work is tracked in side issues and the linked TODOs and REVIEWS.  Another mention above for why this was blocking publication was:

> until we decide how QA should test these nontrivial changes (isolated test vs rolling testing into an upcoming dev or RC test).

I'll do as many of the TODOs and REVIEW comments as I can, but will likely need help from @zepumph on some of them, and it would be good to discuss QA with @zepumph.

**Comment by samreid** [comment]
All TODOs have been addressed or moved to other issues.  The last topic for this issue is how it should be tested.  Some options:

1. (lazy) We alert QA to these changes the next time they are dev or RC testing SHAs that contain these changes.
2. (eager) We create a dedicated version and request up-front testing before other SHAs from master are tested.

The lazy approach may be sufficient in this case, but we have discussed moving toward the eager approach for certain things.  @zepumph and @ariel-phet what do you recommend?

**Comment by zepumph** [comment]
I think part of what leans me towards the lazy approach is the difficulty in pin pointing exactly what may have regressed. It may be easier to test eagerly in cases where you can give a runnable to QA and have them go at it. For this issue I would worry about (in no order):

* Type doc changing
* Methods from the wrapper no longer working
* A bug in IOType that causes an API change
* Something that may not have been converted correctly in a specific sim. Hopefully this would get caught on CT, but in the worst case a sim may have a silent state-related bug that.

Nothing above feels like something that QA can tackle efficiently. Likely CT took us 80% there and a code review can take us the rest. What do you think @samreid?

**Comment by ariel-phet** [comment]
@samreid @zepumph perhaps a bit of both - considering https://github.com/phetsims/QA/issues/519 is high on the testing list currently (EFAC will come first when the next RC is ready, but GAO is actively being tested) - perhaps after some review, GAO will be ready for its next dev test soon and can include these changes with QA alerted. That should help vet the new approach I would think.

**Comment by samreid** [comment]
Other issues such as
https://github.com/phetsims/tandem/issues/212
https://github.com/phetsims/phet-io/issues/1709 
https://github.com/phetsims/tandem/issues/213

Should probably be completed before we start testing.  They will introduce new complexities that would require further tests.

**Comment by samreid** [comment]
I opened a dedicated issue about testing and review (which will span the scope of more than this issue).  All work complete or tracked elsewhere, closing.
